src/lib.rs

```diff
@@ -139,123 +135,190 @@ pub struct Options {
-        .ok_or(failure::err_msg(format!(
-            "index {} is out of bounds of the code section",
-            n - num_imports
-        )))?;
+    module.funcs.par_iter_local_mut().for_each(|(id, func)| {
+        // Don't bother transforming functions that we are snipping.
+        if to_snip.contains(&id) {
+            return;
+        }
+
+        let mut entry = func.entry_block();
+        let v = &mut Replacer { func, to_snip };
+        v.visit_block_id_mut(&mut entry);
+    });
+}
 
-    *body.code_mut().elements_mut() = vec![elements::Instruction::Unreachable, elements::Instruction::End];
+fn unexport_snipped_functions(module: &mut walrus::Module, to_snip: &HashSet<walrus::FunctionId>) {
+    let exports_to_snip: HashSet<walrus::ExportId> = module
+        .exports
+        .iter()
+        .filter_map(|e| match e.item {
+            walrus::ExportItem::Function(f) if to_snip.contains(&f) => Some(e.id()),
+            _ => None,
+        })
+        .collect();
 
-    Ok(())
+    for e in exports_to_snip {
+        module.exports.delete(e);
+    }
 }
 
-trait NamesSectionNames {
-    fn names_section_names(&self) -> Option<elements::NameMap>;
+fn unimport_snipped_functions(module: &mut walrus::Module, to_snip: &HashSet<walrus::FunctionId>) {
+    let imports_to_snip: HashSet<walrus::ImportId> = module
+        .imports
+        .iter()
+        .filter_map(|i| match i.kind {
+            walrus::ImportKind::Function(f) if to_snip.contains(&f) => Some(i.id()),
+            _ => None,
+        })
+        .collect();
+
+    for i in imports_to_snip {
+        module.imports.delete(i);
+    }
 }
 
-impl NamesSectionNames for elements::Module {
-    fn names_section_names(&self) -> Option<elements::NameMap> {
-        for section in self.sections() {
-            if let &elements::Section::Name(elements::NameSection::Function(ref name_section)) =
-                section
-            {
-                return Some(name_section.names().clone());
-            }
+fn snip_table_elements(module: &mut walrus::Module, to_snip: &HashSet<walrus::FunctionId>) {
+    let mut builder = walrus::FunctionBuilder::new();
+    let unreachable = builder.unreachable();
+    let ty = module.types.add(&[], &[]);
+    let unreachable_func = builder.finish(ty, vec![], vec![unreachable], module);
+
+    for t in module.tables.iter_mut() {
+        if let walrus::TableKind::Function(ref mut ft) = t.kind {
+            ft.elements
+                .par_iter_mut()
+                .filter(|el| match el {
+                    Some(f) if to_snip.contains(f) => true,
+                    _ => false,
+                })
+                .for_each(|el| {
+                    *el = None;
+                });
+
+            ft.relative_elements
+                .par_iter_mut()
+                .flat_map(|(_, elems)| elems.par_iter_mut().filter(|f| to_snip.contains(f)))
+                .for_each(|el| {
+                    *el = unreachable_func;
+                });
```

Comment:
I think this isn't right because the signatures can change so existing calls will get a runtime validation error instead of an unreachable? Maybe that's ok though...

Reply:
I was thinking that a panic is a panic, but the error message will probably be a bit misleading, so I'll fix this.

Reply:
Fixed